docs: Configuring the Helm chart and storage (#968)
redpanda-data/documentation: docs/platform/manage-kubernetes/configure-storage.mdx

---
title: Configure Storage in the Helm Chart
tags:

>   - Helm

Author: ```suggestion
  - Kubernetes
```

>   - Helm configuration

Author: ```suggestion
```

---

<head>
    <meta name="title" content="Configure Storage in the Helm Chart | Redpanda Docs"/>
    <meta name="description" content="Configure the Helm chart to use emptyDir volumes, hostPath volumes, or PersistentVolumes."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Redpanda brokers must store their data on disk. You can customize the Helm chart to use the following storage types:

- [emptyDir](#emptydir-volumes)
- [hostPath](#hostpath-volumes)
- [PersistentVolume](#persistentvolume)

By default, the Redpanda Helm chart uses the default StorageClass in a Kubernetes cluster to create one PersistentVolumeClaim for each Pod that runs a Redpanda broker. The default StorageClass in your Kubernetes cluster depends on the Kubernetes platform that you are using.

## Prerequisites

If you're configuring Redpanda for production, make sure to review the [storage best practices](../../deploy-kubernetes/kubernetes-best-practices#storage).

> ## Use emptyDir volumes

Reviewer: ![image](https://user-images.githubusercontent.com/952828/211680766-3b974793-44bc-44da-9377-c5e6ad6fec47.png)

Arranged in this order, it appears that using emptyDir is preferred. Shouldn't we put the default at the top?

An emptyDir volume is first created when a Pod is assigned to a node, and the volume exists as long as the Pod is running on that node.
For details about emptyDir volumes, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir).

To store Redpanda data in emptyDir volumes,
set the `storage.hostPath` configuration to an empty string (`""`),
and set `storage.persistentVolume.enabled` to `false`.

<Tabs groupId="helm-config">
<TabItem value="values" label="--values">

```yaml title="emptydir.yaml"
storage:
  hostPath: ""
  persistentVolume:
    enabled: false
```

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
  --values emptydir.yaml
```

</TabItem>
<TabItem value="flags" label="--set">

```bash
helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
  --set storage.persistentVolume.enabled=false
```

</TabItem>
</Tabs>

:::warning Use only for development and testing
When a Pod is removed from a node for any reason, the data in the emptyDir volume is deleted permanently.

> :::

Reviewer: Should the warning be right under the section title? I'd hate for someone to call support because they lost all their data only to realize they didn't scroll down far enough (seen it happen).

  ```

  </TabItem>
  </Tabs>

</TabItem>
</Tabs>

### Resizing Persistent Volumes

To give Redpanda brokers more storage, you can expand the size of PersistentVolumes.

1. Make sure that your StorageClass is capable of volume expansions. For a list of volumes that support volume expansion, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/storage-classes/#allow-volume-expansion).

1. Increase the value of the `storage.persistentVolume.size` configuration:

  <Tabs groupId="helm-config">
  <TabItem value="values" label="--values">

  ```yaml title="persistentvolume-size.yaml"
  storage:
    persistentVolume:
      enabled: true
      size: <custom-size>Gi
  ```

  ```bash
  helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
    --values persistentvolume-size.yaml
  ```

  </TabItem>
  <TabItem value="flags" label="--set">

  ```bash
  helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
    --set storage.persistentVolume.enabled=true \
    --set storage.persistentVolume.size=<custom-size>Gi
  ```

  </TabItem>
  </Tabs>

### Deleting PersistentVolumeClaims

> PersistentVolumesClaims are not deleted when Redpanda brokers are removed from a cluster.

Reviewer: Do we want to add this?

```suggestion
To prevent accidental loss of data, PersistentVolumesClaims are not deleted when Redpanda brokers are removed from a cluster.
```

## Use PersistentVolumes

A PersistentVolume is storage in the cluster that has been provisioned by an administrator or dynamically provisioned using StorageClasses.
For details about PersistentVolumes, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).

You can configure the Helm chart to use PersistentVolumes with a static provisioner or a dynamic provisioner.
Redpanda recommends using a StorageClass with a dynamic provisioner. See the [best practices](../../deploy-kubernetes/kubernetes-best-practices).

- `storage.persistentVolume.storageClass` sets the StorageClass that the Helm chart configures for each PersistentVolumeClaim.
- `storage.persistentVolume.size` sets the amount of resources that is requested by each PersistentVolumeClaim.
- `storage.persistentVolume.labels` sets the labels to apply to each PersistentVolumeClaim.
- `storage.persistentVolume.annotations` sets the annotations to apply to each PersistentVolumeClaim.

<Tabs groupId="provisioners">
<TabItem value="dynamic" label="Dynamic provisioners">

> When you use a dynamic provisioner, it is responsible for creating a PersistentVolume on demand for each Redpanda broker.

Reviewer 2: ```suggestion
When you use a dynamic provisioner, it creates a PersistentVolume on demand for each Redpanda broker.
```
